live-debugger-ffi/src/data.rs

```diff
@@ -0,0 +1,297 @@
+        std::mem::forget(tags);
+        new
+    }
+}
+
+#[no_mangle]
+extern "C" fn drop_span_decoration_probe(_: SpanDecorationProbe) {}
+
+impl<'a> Drop for SpanDecorationProbe<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            let tags = Vec::from_raw_parts(
+                self.span_tags as *mut SpanProbeTag,
+                self.span_tags_num,
+                self.span_tags_num,
+            );
+            let num_conditions = tags.iter().filter(|p| p.next_condition).count();
+            _ = Vec::from_raw_parts(
+                self.conditions as *mut ProbeCondition,
+                num_conditions,
+                num_conditions,
+            );
+        };
+    }
+}
+
+#[repr(C)]
+pub enum ProbeType<'a> {
+    Metric(MetricProbe<'a>),
+    Log(LogProbe<'a>),
+    Span,
+    SpanDecoration(SpanDecorationProbe<'a>),
+}
+
+impl<'a> From<&'a datadog_live_debugger::ProbeType> for ProbeType<'a> {
+    fn from(from: &'a datadog_live_debugger::ProbeType) -> Self {
+        match from {
+            datadog_live_debugger::ProbeType::Metric(metric) => ProbeType::Metric(metric.into()),
+            datadog_live_debugger::ProbeType::Log(log) => ProbeType::Log(log.into()),
+            datadog_live_debugger::ProbeType::Span(_) => ProbeType::Span,
+            datadog_live_debugger::ProbeType::SpanDecoration(span_decoration) => {
+                ProbeType::SpanDecoration(span_decoration.into())
+            }
+        }
+    }
+}
+
+#[repr(C)]
+pub struct ProbeTarget<'a> {
+    pub type_name: Option<CharSlice<'a>>,
+    pub method_name: Option<CharSlice<'a>>,
+    pub source_file: Option<CharSlice<'a>>,
+    pub signature: Option<CharSlice<'a>>,
+    pub lines: CharSliceVec<'a>,
+    pub in_body_location: InBodyLocation,
+}
+
+impl<'a> From<&'a datadog_live_debugger::ProbeTarget> for ProbeTarget<'a> {
+    fn from(from: &'a datadog_live_debugger::ProbeTarget) -> Self {
+        ProbeTarget {
+            type_name: from.type_name.as_ref().map(|s| s.as_str().into()).into(),
+            method_name: from.method_name.as_ref().map(|s| s.as_str().into()).into(),
+            source_file: from.source_file.as_ref().map(|s| s.as_str().into()).into(),
+            signature: from.signature.as_ref().map(|s| s.as_str().into()).into(),
+            lines: (&from.lines).into(),
+            in_body_location: from.in_body_location,
+        }
+    }
+}
+
+#[repr(C)]
+pub struct Probe<'a> {
+    pub id: CharSlice<'a>,
+    pub version: u64,
+    pub language: Option<CharSlice<'a>>,
+    pub tags: CharSliceVec<'a>,
+    pub target: ProbeTarget<'a>, // "where" is rust keyword
+    pub evaluate_at: EvaluateAt,
+    pub probe: ProbeType<'a>,
+}
+
+impl<'a> From<&'a datadog_live_debugger::Probe> for Probe<'a> {
+    fn from(from: &'a datadog_live_debugger::Probe) -> Self {
+        Probe {
+            id: from.id.as_str().into(),
+            version: from.version,
+            language: from.language.as_ref().map(|s| s.as_str().into()).into(),
```

Comment:
Is there a note/explanation somewhere for what this is for? Are there multiple string types being used?

Reply:
Not sure what you are asking? Why I'm collecting the language? What this particular line of rust does? Or just asking what I'm doing to convert an Option<String> into an Option<CharSlice>?

Reply:
Apologies for being unclear. I was asking why `language: from.language` would be insufficient, for my general education on how rust works.

Reply:
Ah, yeah, just the conversion of the value within the Option to something FFI understands.